BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -0,0 +1,377 @@
+    CAN_InitStruct.CAN_Prescaler = 16;
+    CAN_Init(CAN3, &CAN_InitStruct);
+
+    /* CAN filter init */
+    CAN_FilterInitStruct.CAN_FilterNumber = 0;
+    CAN_FilterInitStruct.CAN_FilterMode = CAN_FilterMode_IdMask;
+    CAN_FilterInitStruct.CAN_FilterScale = CAN_FilterScale_32bit;
+    CAN_FilterInitStruct.CAN_FilterIdHigh = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterIdLow = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterMaskIdHigh = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterMaskIdLow = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterFIFOAssignment = 0;
+    CAN_FilterInitStruct.CAN_FilterActivation = ENABLE;
+    CAN_FilterInit(CAN3, &CAN_FilterInitStruct);
+
+    //CAN_SlaveStartBank(CAN1, 0);
+
+    /* Transmit Structure preparation */
+    gTxMessage.ExtId = 0x5;
+    gTxMessage.RTR = CAN_RTR_DATA;
+    gTxMessage.IDE = CAN_ID_STD;
+    gTxMessage.DLC = 1;
+
+    /* Receive Structure preparation */
+    gRxMessage.StdId = 0x00;
+    gRxMessage.ExtId = 0x00;
+    gRxMessage.IDE = CAN_ID_STD;
+    gRxMessage.DLC = 0;
+    gRxMessage.FMI = 0;
+
+    /* Enable FIFO 0 message pending Interrupt */
+    CAN_ITConfig(CAN3, CAN_IT_FMP0, ENABLE);
+
+    // Enable Rx interrupts
+    NVIC_InitStruct.NVIC_IRQChannel = CAN3_RX0_IRQn;
+    NVIC_InitStruct.NVIC_IRQChannelPreemptionPriority = 0x00;
+    NVIC_InitStruct.NVIC_IRQChannelSubPriority = 0x00;
+    NVIC_InitStruct.NVIC_IRQChannelCmd = ENABLE;
+    NVIC_Init(&NVIC_InitStruct);	
+
+}
+
+/**
+ * @brief   Transmits the data onto the CAN bus with the specified id
+ * @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
+ * @param   data : data to be transmitted. The max is 8 bytes.
+ * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
+ * @return  ERROR if module was unable to transmit the data onto the CAN bus. SUCCESS indicates data was transmitted.
+ */
+ErrorStatus BSP_CAN_Write(CAN_t bus, uint32_t id, uint8_t data[8], uint8_t length) {
+    
+    gTxMessage.StdId = id;
+    gTxMessage.DLC = length;
+	for(int i = 0; i < length; i++){
+        gTxMessage.Data[i] = data[i];
+    }
+	
+    ErrorStatus retVal;
+    
+    if(bus == CAN_1){
+        retVal = (ErrorStatus) (CAN_Transmit(CAN1, &gTxMessage) != 0);
+    }else{
+        retVal = (ErrorStatus) (CAN_Transmit(CAN3, &gTxMessage) != 0);
+    }
+
+    return retVal;
+}
+
+/**
+ * @brief   Gets the data that was received from the CAN bus.
+ * @note    Non-blocking statement
+ * @pre     The data parameter must be at least 8 bytes or hardfault may occur.
+ * @param   id : pointer to store id of the message that was received.
+ * @param   data : pointer to store data that was received. Must be 8bytes or bigger.
+ * @return  ERROR if nothing was received so ignore id and data that was received. SUCCESS indicates data was received and stored.
+ */
+ErrorStatus BSP_CAN_Read(CAN_t bus, uint32_t *id, uint8_t *data) {
```

Comment:
We don't appear to be using the `bus` parameter at all here. Is it still necessary? If so, we should separate out the rx queue into two (one for each bus). If not, we should remove the parameter from the argument list.